docs/ENSEMBLES.md

```diff
@@ -0,0 +1,57 @@
+# Design blueprint: ensembles
+
+We want testground to support running a test plan with instances built against
+different versions sets. For example, running the DHT test plan with 200
+instances split into groups:
+
+- Group A: 50 instances are built against go-libp2p-kad-dht/v0.3.0,
+  go-libp2p-swarm/v0.3.0
+- Group B: 50 instances are built against go-libp2p-kad-dht/v0.3.0,
+  go-libp2p-swarm/v0.4.0
+- Group C: 50 instances are built against go-libp2p-kad-dht/v0.4.0,
+  go-libp2p-swarm/v0.3.0
+- Group D: 50 instances are built against go-libp2p-kad-dht/v0.4.0,
+  go-libp2p-swarm/v0.4.0
+
+Testground has been designed to offer the primitives to compose such **test
+ensembles**. We coin the term **test ensemble** to refer to a set of test
+instances, potentially built against different dependency graphs, potentially
+running with different parameters, but orchestrated by the same runner, and
+participating under the same *run id*.
+
+Thus, all instances comprising the ensemble have visibility over one another,
+yet they are configured differently, and such variance is totally opaque.
+
+As long as the DHT plan can be compiled against all upstream versions of the
+dependency graph being tested, scheduling the above ensemble would entail:
+
+1. Building the plan 4 times, passing in the appropriate dependency sets (via
+   the CLI, this is accomplished with the `--dep` param) each time. We now have
+   4 build artifacts, each with its own ID.
+2. Running 50 instances of each build artifact, enlisting them in the same run
+   ID, providing total instance count=200. This results in all instances
+   attaching to the same sync subtree, and awaiting for 200 instances total to
+   appear.
+
+As the test scenario runs, each instance outputs metrics and assets. The
+challenge now becomes how do we differentiate the metrics generated by each
+dependency graph / test parameter set.
+
+In practice, we want to answer questions like: "how many dials per second did
```

Comment:
Just to be clear, when you say `how many dials per second did version X generate`, you mean a specific library version, not a specific group, right?

1. If we care about stats from a given Group, we just need to tag the binary/docker image and sent stats with a given tag.
2. If we care about stats from `go-libp2p-swarm/v0.3.0`, then we need to filter for results from both Group A and Group C, which is indeed more complicated than 1. and we need the dep graph tracking, as described below.

Reply:
I'm talking about concrete upstream dependency versions. IMO groups are a higher-level construct. We'll want to record which groups ran with a particular test run, but when it comes to a comparative analysis, I'm not entirely sure they'll provide much value. The kind of question I'd like to answer is: how did go-libp2p-swarm/v0.2.0 perform against 0.1.0?